lib/CL/clCreateBuiltinKernelWithAttributes.cc

```diff
@@ -0,0 +1,480 @@
+
+  // TBC: 4D+ tensor could be supported by treating the additional
+  //      dimensions as batch dimensions - but it might not be
+  //      worthwhile due the extra work to support them and processing
+  //      overhead they may impose.
+  if (A.rank() > 3)
+    return LogError(CL_DBK_INVALID_ATTRIBUTE,
+                    "Unsupported high-degree tensors.");
+
+  if (A.rank() != B.rank() || B.rank() != CO.rank())
+    // TODO: Should we have something like CL_DBK_INVALID_TENSOR_SHAPE?
+    return LogError(CL_DBK_INVALID_ATTRIBUTE, "Rank mismatch.");
+
+  if (CIOpt && !CIOpt->shapeEquals(CO))
+    return LogError(CL_DBK_INVALID_ATTRIBUTE,
+                    "Tensor shape mismatch between c_in and c_out.");
+
+  // FIXME: check tensor shapes are correct respect to the transpose
+  //        configurations.
+
+  size_t BatchDims = A.rank() - 2;
+
+  // CO[b][m][n] = sigma_over_m_n_k(A[b][m][k] * B[b][k][n]) + CI[b][m][n].
+  auto Am = A[BatchDims + 0];
+  auto Ak = A[BatchDims + 1];
+  auto Bk = B[BatchDims + 0];
+  auto Bn = B[BatchDims + 1];
+  auto COm = CO[BatchDims + 0];
+  auto COn = CO[BatchDims + 1];
+
+  if (Ak != Bk || Am != COm || Bn != COn)
+    // TODO: Should have more descriptive error code? Or would it be better
+    //       to have error logging like the cl_program has for building?
+    return LogError(CL_DBK_INVALID_ATTRIBUTE, "Matrix shape mismatch.");
+
+  // Check batch dimensions match.
+  size_t BatchSize = A.rank() == 3 ? A[0] : 1;
+  if (BatchSize > 1 && (BatchSize != B[0] || B[0] != CO[0]))
+    return LogError(CL_DBK_INVALID_ATTRIBUTE, "Batch size mismatch.");
+
+  if (BatchSize > 1 && CIOpt && (*CIOpt)[0] != CO[0])
+    return LogError(CL_DBK_INVALID_ATTRIBUTE, "Batch size mismatch.");
+
+  if (A.dtype() != B.dtype())
+    return LogError(CL_DBK_INVALID_ATTRIBUTE,
+                    "dtype mismatch between A and B.");
+
+  if (CIOpt && CIOpt->dtype() != CO.dtype())
+    return LogError(CL_DBK_INVALID_ATTRIBUTE,
+                    "dtype mismatch between input and output C.");
+
+  // TODO: We probably need to have support for mixed input/output
+  // precisions to be able to fit results of large, low precision input
+  // matrices. precision inputs. E.g.
+  //
+  //  * i8 x i8   --> i32
+  //  * f16 x f16 --> f32
+  if (A.dtype() != CO.dtype())
+    return LogError(CL_DBK_INVALID_ATTRIBUTE, "Unsupported I/O dtype");
+
+  // TODO: extend support for other data types.
+  if (A.dtype() != CL_TENSOR_FLOAT)
+    return LogError(CL_DBK_UNAVAILABLE, "Unimplemented dtype support.");
+
+  // TODO: check validity of data layouts of the tensors. Now assume
+  // they are correct and they are using BLAS-like layout.
+
+  float Alpha = 1.0f, Beta = 0.0f;
+  if (AlphaAttr)
+    std::memcpy(&Alpha, AlphaAttr, sizeof(float));
+  if (BetaAttr)
+    std::memcpy(&Beta, BetaAttr, sizeof(float));
+
+  // libxsmm does not support arbitrary alpha and beta (for now).
```

Comment:
The DBK implementations using libxsmm should go to lib/devices/basic where we can pick them up for the different CPU devices.